Allow 'firebase deploy' from any directory within app.

TamalSaha opened the pull request:
+cc: @robertdimarco

jwngr commented:
Re-assigning to Chris since he knows this code much better than I do.

robertdimarco commented:
@TamalSaha The automated tests are failing; check out the results by using the `Details` links above. You can also run that suite locally with `gulp`.

TamalSaha commented:
@robertdimarco Please check again.

drtriumph commented on lib/app.js:
```diff
@@ -294,13 +294,20 @@ function uploadSite(settings, directoryRef, argv) {
-    console.log(chalk.red('Initialization Error') + ' - Directory not ' +
+  var curDir = process.cwd();
+  var settingsFile;
+  while(!fs.existsSync((settingsFile = (curDir + '/firebase.json')))) {
```
While this is probably correct and concise, it's rather terse. We tend to avoid assignment in while loop definitions, and I tend to be a proponent of recursive functions in places like this as they can be easier to reason about

drtriumph replied:
also, in addition, we'd put a space after control keywords like `while`, `if`, `catch` etc

drtriumph replied:
also, I think the `path.resolve()` is significant in that it turns unix-style paths like `./firebase.json` and converts then to windows paths if relevant: `.\\firebase.json`

drtriumph commented on lib/app.js:
```diff
@@ -642,17 +649,20 @@ module.exports = {
-          if (defaultSettings.hasOwnProperty(i) &&
+          if (i === 'public') {
+            if (settings.hasOwnProperty('public')) {
+              settings.public = path.resolve(defaultSettings.public, settings.public);
```
We need to still make sure they don't specify a public directory that is above the directory that the `firebase.json` is in - I think that's still an important constraint to prevent people accidentally uploading their home folder etc.

drtriumph commented:
I made some comments, back to you

TamalSaha commented:
@drtriumph I have addressed your comments except the comment about using recursion. I like to avoid recursion in JS in general, since JS does not perform tail call optimization.

drtriumph commented on lib/app.js:
```diff
@@ -642,15 +649,23 @@ module.exports = {
-          if (defaultSettings.hasOwnProperty(i) &&
+          if (i === 'public') {
+            if (settings.hasOwnProperty('public')) {
+              settings.public = path.resolve(defaultSettings.public, settings.public);
```
The use of `defaultSettings.public` here is confusing - to make it more readable we should rename `curDir` to something more descriptive like `projectRoot` of `projectRootDir` and then we could use that directly here

drtriumph commented on lib/app.js:
```diff
@@ -642,15 +649,23 @@ module.exports = {
         }
-        if (path.relative('.', settings.public).match(/^\.\./)) {
+
+        if (path.relative(defaultSettings.public, settings.public).match(/^\.\./)) {
```
Again, this should be `projectRootDir` to make it clear what we're doing

drtriumph commented on lib/app.js:
```diff
@@ -642,15 +649,23 @@ module.exports = {
         for (var i in defaultSettings) {
-          if (defaultSettings.hasOwnProperty(i) &&
+          if (i === 'public') {
+            if (settings.hasOwnProperty('public')) {
```
We should `&&` these two `if()` statements together so they can share the same `else`

drtriumph commented on lib/app.js:
```diff
@@ -294,13 +294,20 @@ function uploadSite(settings, directoryRef, argv) {
-  if (!fs.existsSync(settingsFile)) {
-    console.log(chalk.red('Initialization Error') + ' - Directory not ' +
+  var curDir = process.cwd();
+  while (!fs.existsSync(path.resolve(curDir, './firebase.json'))) {
```
This is more readable now and acceptable, thank you. As an aside - it's unlikely that this would have been called more than 10 times deep ever, so the lack of tail recursion in JavaScript would have had negligible effect

drtriumph commented:
A few more comments, back to you. Also, I want to see the functionality demonstrated on a Mac, Linux and a Windows computer

TamalSaha commented:
@drtriumph I have addressed your comments. I can demonstrate you the new behavior in Win/Lin/Mac.

drtriumph commented:
Ok, what I can see looks good, I'll take a look at the cross-platform functionality in the morning
